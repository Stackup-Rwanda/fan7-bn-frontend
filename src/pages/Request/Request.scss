.request-page {
  padding: 1.25em;
}
.big-container{
  position: fixed;  
  width: 100%;  
  height: 100%;  
  left: 0;  
  right: 0;  
  bottom: 0;  
  margin: auto;  
  background: #ECECEC;
  display: block;
  font-family: 'Poppins', sans-serif;
  font-size: 13px;

}

.sub-container{
  display: block;
  width: 79%;
  height: 100%;
  margin-left: 19%;
  margin-top: 7%;

  .mainButton{
    width: 100%;
    background: white;
    margin-left: 0%;

    .buton{
      width: 20%;
      margin-left: 1%;
      margin-bottom: 10px;
    }  
  }
  
}
.line{
  opacity: 0.3;
  border: 1px solid #C4C4C4;
}
.emptyDiv{
  width: 100%;
  height: 20px;
  background: #FFFFFF;
}
.search{
  width: 100%;
  height: 30px;
  background: #FFFFFF;
  &-box {    
    position: absolute;
    display: flex;
    justify-content: space-between;
    width: 150px;
    height: 25px;
    left: 80%;
    margin-top: -15px;
  }
  .page-title {
    padding: 5px 0 0 20px;
    display: inline-block;
  }
}
