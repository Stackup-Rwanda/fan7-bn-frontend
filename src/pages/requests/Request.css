.reqContainer {
  position: fixed;
  width: 100%;
  height: 100%;
  left: 0;
  right: 0;
  bottom: 0;
  margin: auto;
  background: #ECECEC;
  display: block;
  font-family: 'Poppins', sans-serif;
  font-size: 13px;
}

.reqContainer_mini-container {
  display: block;
  background: white;
  width: 79%;
  height: 80%;
  margin-left: 19%;
  margin-top: 7%;
}

.reqContainer_mini-container_mainButton {
  width: 100%;
  margin-left: 1%;
  padding-bottom: 10px;
}

.reqContainer_mini-container_mainButton .buton {
  width: 20%;
  margin-left: 0%;
}

.reqContainer_mini-container_line {
  opacity: 0.3;
  border: 1px solid #C4C4C4;
}

.reqContainer_mini-container_search {
  width: 100%;
  padding-bottom: 10px;
}

.reqContainer_mini-container_search_page-title {
  padding: 10px 0 0 20px;
  display: inline-block;
}

.reqContainer_mini-container_mainTable {
  width: 100%;
  height: 70%;
}

.reqContainer_sidebar {
  height: 100%;
  width: 17%;
  background: #08233C;
  position: fixed;
  left: 0;
  z-index: 1;
  top: 0;
}

.reqContainer_navbar {
  width: 100%;
  height: 10%;
  background: white;
  position: fixed;
}
/*# sourceMappingURL=Request.css.map */