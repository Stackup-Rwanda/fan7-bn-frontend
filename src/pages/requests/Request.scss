
.reqContainer{
    position: fixed;  
    width: 100%;  
    height: 100%;  
    left: 0;  
    right: 0;  
    bottom: 0;  
    margin: auto;  
    background: #ECECEC;
    display: block;
    font-family: 'Poppins', sans-serif;
    font-size: 13px;

    &_mini-container{
        display: block;
        background: white;
        width: 79%;
        height: 80%;
        margin-left: 19%;
        margin-top: 7%;
    
&_mainButton{
    width: 100%;
    margin-left: 1%;
   padding-bottom: 10px;


.buton{
    width: 20%;
    margin-left: 0%;

}
}

&_line{
    opacity: 0.3;
    border: 1px solid #C4C4C4;
}
&_search{
    width: 100%;
    padding-bottom: 10px;

    &_page-title{
        padding: 10px 0 0 20px;
        display: inline-block;
    }
}
&_mainTable{
    width: 100%;
    height: 70%;
}
    }
    &_sidebar{
        height: 100%;
        width: 17%;
        background: #08233C;
        position: fixed;
        left: 0;
        z-index: 1;
        top: 0; 
    
    }
    &_navbar{
        width: 100%;
        height: 10%;
        background: white;
        position: fixed;
    
    
    }
}

