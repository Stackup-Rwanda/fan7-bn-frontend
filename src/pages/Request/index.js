import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import jwtDecode from 'jwt-decode';
import Table from '../../components/Table';
import {
  create_one_way_trip,
  create_return_trip,
  create_multi_city_trip,
  get_requests,
  editTripRequest

} from '../../store/modules/requests/userRequests/actions';
import { getTripRequests } from '../../store/modules/request/view/actions';
import {
  selectLoading,
  selectCount,
  selectRequests,
} from '../../store/modules/request/view/selectors';
import Popup from '../../components/popup/popup';
import AuthService from '../../utils/AuthService';
import './Request.scss';
import DynamicDashboard from '../../components/DynamicDashboard/Dashboard';
import ServerSidePagination from '../../components/ServerPagination/ServerSidePagination';
import { requesterDashboard, managerDashboard } from '../../assets/sidebar';
import profileImg from '../../assets/images/icons8-user-30.png';
import InputField from '../../components/InputField';
import Button from '../../components/Button';

class Request extends Component {
  constructor(props) {
    super(props);
    this.initialState = {
      showPopup: false,
      requests: [],
      data: null
    };

    this.state = this.initialState;
    this.handleAction = this.handleAction.bind(this);
    this.onChangePage = this.onChangePage.bind(this);
  }

  componentDidMount() {
    const { dispatch } = this.props;
    dispatch(getTripRequests(1, 5));
  }

  handleAction(row) {
    console.log(row);
  }
  onChangePage(page, limit) {
    const { dispatch } = this.props;
    dispatch(getTripRequests(page, limit));
  }
  togglePopup() {
    this.setState({
      showPopup: !this.state.showPopup,
      data: null
    });
  }
  toggleEditPopup(data) {
    this.setState({
      showPopup: !this.state.showPopup,
      data
    });
  }
  oneway(data) {
    const { dispatch } = this.props;
    dispatch(create_one_way_trip(data))
  }
  returnTrip(data) {
    const { dispatch } = this.props;
    dispatch(create_return_trip(data))
  }
  multicity(data) {
    const { dispatch } = this.props;
    dispatch(create_multi_city_trip(data))
  }
  handleTripRequestUpdate(data, id) {
const { dispatch } = this.props;
    dispatch(editTripRequest(data, id))
  }

  render() {
    const managerCols = [
      { header: 'Requester', name: 'requester' },
      { header: 'Trip Reason', name: 'reason' },
      { header: 'Destination(s)', name: 'destination' },
      { header: 'Travel Date', name: 'travel_date' },
      { header: 'Return Date', name: 'returnDate' },
      { header: 'Type', name: 'type' },
      { header: 'Status', name: 'status' },
    ];
    const requesterCols = [
      { header: 'Trip Reason', name: 'reason' },
      { header: 'Destination(s)', name: 'destination' },
      { header: 'Travel Date', name: 'travel_date' },
      { header: 'Return Date', name: 'returnDate' },
      { header: 'Passport Name', name: 'passportName' },
      { header: 'Passport Number', name: 'passportNumber' },
      { header: 'Type', name: 'type' },
      { header: 'Status', name: 'status' },
    ];
    const { loading, count, requests } = this.props;
    const token = AuthService.getToken();
    const { role } = !!token ? jwtDecode(token) : { role: '' };
    const board = role === 'requester' ? requesterDashboard : managerDashboard;
    const createBtn = role === 'requester' ?
      <div className="mainButton">
        <Button
          type="submit"
          className="btn buton"
          value=" + Create Trip Request"
          onClick={this.togglePopup.bind(this)}
        />


        {this.state.showPopup ?
          <Popup
            oneWay={this.oneway.bind(this)}
            returntrip={this.returnTrip.bind(this)}
            multiCity={this.multicity.bind(this)}
            closePopup={this.togglePopup.bind(this)}
            data={this.state.data}
            update={this.handleTripRequestUpdate.bind(this)}
          />
          : null
        }
      </div>
      : <div className="emptyDiv"></div>;

    return (
      <div className="big-container">
        <div className="sub-container">
          {createBtn}
          <hr className="line"></hr>
          <div className="search">
          <label className="page-title">Records per page</label> 
          <div className="search-box">
          <label className="">Search:</label> 
          <InputField
          type="text"
          name="search"
          className="search-input"
          />
          </div>
          </div>
          <hr className="line"></hr>
          <DynamicDashboard
          properties={board}
          profile={profileImg}
          />
          <Table
            cols={(role === 'requester' && requesterCols) || (role === 'manager' && managerCols) || []}
            data={requests}
            loading={loading}
            actions={true}
            handleAction={this.handleAction}
            handleEdit={role === 'requester' && this.toggleEditPopup.bind(this)}
          />
            {/* Pagination start*/}
            <ServerSidePagination
            handleChangePage={this.onChangePage}
            totalRows={count}
            />
            {/* Pagination end */}
          </div>
      </div>
    );
  }
}

Request.propTypes = {
  dispatch: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
  count: PropTypes.number.isRequired,
  requests: PropTypes.arrayOf().isRequired,
};

const select = ({ viewRequests }) => ({
  loading: selectLoading(viewRequests),
  count: selectCount(viewRequests),
  requests: selectRequests(viewRequests),
});

export default connect(select)(Request);
