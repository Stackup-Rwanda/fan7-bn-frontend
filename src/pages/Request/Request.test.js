/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { mount, shallow } from 'enzyme';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { BrowserRouter as Router } from 'react-router-dom';
import Request from './';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const history = {
  push: jest.fn(),
  replace: jest.fn(),
  createHref: jest.fn(),
};

describe('Request Component Test', () => {
  it('should return true if component exists', () => {
    const initialState = {};
    const store = mockStore(initialState);
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <Request />
        </Router>
      </Provider>
    );

    expect(wrapper.find('.big-container').exists()).toBe(true);
  });
});

describe('<Request /> shallow rendering tests', () => {
  const initialState = {};
  const store = mockStore(initialState);

  it('matches the snapshot', () => {
    const tree = shallow(<Request store={store} />);
    expect(tree).toMatchSnapshot();
  });
});
