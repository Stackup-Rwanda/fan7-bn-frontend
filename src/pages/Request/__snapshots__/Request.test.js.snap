// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Request /> shallow rendering tests matches the snapshot 1`] = `
<ContextProvider
  value={null}
>
  <Request
    count={0}
    dispatch={[Function]}
    loading={false}
    requests={Array []}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  />
</ContextProvider>
`;
